timesketch/tsctl.py

```diff
@@ -1351,3 +1363,579 @@ def list_config():
+        else:
+            print("  Input is not JSONL, using as CSV...")
+            event_data_bytes = input_content.encode("utf-8")
+
+    elif output_format == "jsonl":
+        if is_likely_jsonl:
+            print("  Input is JSONL, using as is...")
+            event_data_bytes = input_content.encode("utf-8")
+        else:
+            print("  Input is not JSONL, converting CSV to JSONL...")
+            try:
+                csv_input = io.StringIO(input_content)
+                try:
+                    dialect = csv.Sniffer().sniff(csv_input.read(1024))
+                    csv_input.seek(0)
+                    reader = csv.DictReader(csv_input, dialect=dialect)
+                    print(
+                        f"    Detected CSV dialect: delimiter='{dialect.delimiter}'"
+                    )  # pylint: disable=line-too-long
+                except csv.Error:
+                    print(
+                        "    Could not detect CSV dialect, assuming comma delimiter."
+                    )  # pylint: disable=line-too-long
+                    csv_input.seek(0)
+                    reader = csv.DictReader(csv_input)
+
+                output_jsonl = io.StringIO()
+                count = 0
+                for row in reader:
+                    processed_row = {}
+                    for key, value in row.items():
+                        try:
+                            processed_row[key] = int(value)
+                            continue
+                        except (ValueError, TypeError):
+                            pass
+                        try:
+                            processed_row[key] = float(value)
+                            continue
+                        except (ValueError, TypeError):
+                            pass
+                        if isinstance(value, str):
+                            if value.lower() == "true":
+                                processed_row[key] = True
+                                continue
+                            if value.lower() == "false":
+                                processed_row[key] = False
+                                continue
+                        processed_row[key] = value
+
+                    output_jsonl.write(json.dumps(processed_row) + "\n")
+                    count += 1
+
+                event_data_bytes = output_jsonl.getvalue().encode("utf-8")
+                print(f"    Successfully converted {count} CSV rows to JSONL.")
+
+            except Exception as conversion_error:  # pylint: disable=broad-except
+                print(f"  ERROR converting CSV to JSONL: {conversion_error}")
+                print(traceback.format_exc())
+                raise  # Re-raise to stop execution
+
+    return event_data_bytes
+
+
+# Helper function to create the zip archive
+def _create_export_archive(
+    filename: str, metadata: dict, event_data_bytes: bytes, event_filename: str
+):
```

Comment:
Consider using atomic file writing: Instead of writing directly to the final filename, we can write to a temporary file in the same directory. Or create a mktmp directory and write batches of events into separate files. This tmp folder can then be zipped.
If the zip creation is successful, we then atomically replace (`os.replace()`) the temporary file/folder with the final filename. If any error occurs, we clean up the temporary file. This ensures that the target filename is only created or replaced once the entire content is successfully written.

Especially the batched writing of events would reduce the impact on RAM and make it more resilient for large Sketch exports.

Reply:
filed https://github.com/google/timesketch/issues/3415